packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts

```diff
@@ -0,0 +1,412 @@
+            throw new Error(`Failed to write summary tree`);
+        }
+        this.blobTreeDedupCaches = { ...blobTreeDedupCachesLatest };
+        this.lastSummaryProposalHandle = id;
+        return id;
+    }
+
+    private async writeSummaryTreeCore(
+        parentHandle: string | undefined,
+        tree: api.ISummaryTree,
+    ): Promise<{
+            result: ISnapshotResponse,
+            blobTreeDedupCachesLatest: IDedupCaches,
+    }> {
+        // This cache is associated with mapping sha to path for currently generated summary.
+        // We are building these caches from scratch as this will take care of the deleted blobs. The deleted blobs/trees will not come
+        // in these caches and then we will replace the old caches with these new caches, so that they have the correct values.
+        const blobTreeDedupCachesLatest: IDedupCaches = {
+            blobShaToPath: new Map(),
+            pathToBlobSha: new Map(),
+            treesPathToTree: new Map(),
+        };
+        const { snapshotTree, reusedBlobs, blobs } = await this.convertSummaryToSnapshotTree(
+            parentHandle,
+            this.addAppPrefixToSummaryTree(tree),
+            blobTreeDedupCachesLatest,
+            ".app",
+            "",
+            false,
+        );
+        const snapshot: ISnapshotRequest = {
+            entries: snapshotTree.entries!,
+            message: "app",
+            type: SnapshotType.Channel,
+        };
+
+        return getWithRetryForTokenRefresh(async (options) => {
+            const storageToken = await this.getStorageToken(options, "WriteSummaryTree");
+
+            const { url, headers } = getUrlAndHeadersWithAuth(`${this.snapshotUrl}/snapshot`, storageToken);
+            headers["Content-Type"] = "application/json";
+
+            const postBody = JSON.stringify(snapshot);
+
+            return PerformanceEvent.timedExecAsync(this.logger,
+                {
+                    eventName: "uploadSummary",
+                    attempt: options.refresh ? 2 : 1,
+                    hasClaims: !!options.claims,
+                    headers: Object.keys(headers).length !== 0 ? true : undefined,
+                    blobs,
+                    reusedBlobs,
+                    size: postBody.length,
+                },
+                async () => {
+                    const response = await this.epochTracker.fetchAndParseAsJSON<ISnapshotResponse>(
+                        url,
+                        {
+                            body: postBody,
+                            headers,
+                            method: "POST",
+                        },
+                        FetchType.uploadSummary);
+                    return { result: response.content, blobTreeDedupCachesLatest };
+                });
+        });
+    }
+
+    /**
+     * Following are the goals of this function.
+     *  a.) Converts the summary tree to a snapshot/odsp tree to be uploaded. Always upload full snapshot tree.
```

Comment:
We should open an issue to follow up on uploading full snapshot. This is not ideal (but likely Ok for staging).
At a minimum we should observe summarizer stats (we record sizes, including number of nodes).
Can you please double check if we record payload sizes (i.e. stringified result)? If not, we might want to start tracking it.

Any way to make this conditional? I.e. such that if we learn that it makes things worse (i.e. cons of that is bigger than benefits from blob de-dup), it would be great to have simple way to turn it off and figure out next steps.

Reply:
Yes, we record payload sizes. I have added a boolean "allowHandleExpansion". So in case we immediately want to turn it off we will just set it to false.

Reply:
Opened new issue for the same. https://github.com/microsoft/FluidFramework/issues/4817